### ISSUE TITLE
Provision server certificates for controller manager and scheduler

### ISSUE THREAD

**Issue by rifelpet** [kubernetes/kops#12016]
/kind feature

**1. Describe IN DETAIL the feature/behavior/change you would like to see.**

Failing tests on jobs using the release/latest k8s version marker show that we arent providing server certificates to kube-controller-manager and kube-scheduler for exposing their metrics endpoints:

https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/e2e-kops-grid-scenario-aws-cloud-controller-manager/1416404459017211904

This is the k/k PR that introduces a dependency on the secure metrics ports of these components: https://github.com/kubernetes/kubernetes/pull/102050

More discussion here: https://kubernetes.slack.com/archives/C8MKE2G5P/p1626535888459700 

 
**2. Feel free to provide a design supporting your feature request.**

Relevant flags:

```
$ docker run -it --rm --entrypoint=/usr/local/bin/kube-scheduler k8s.gcr.io/kube-scheduler:v1.22.0-beta.2 --help

Secure serving flags:

      --bind-address ip
                The IP address on which to listen for the --secure-port port. The associated interface(s) must be reachable by the rest of the cluster,
                and by CLI/web clients. If blank or an unspecified address (0.0.0.0 or ::), all interfaces will be used. (default 0.0.0.0)
      --cert-dir string
                The directory where the TLS certs are located. If --tls-cert-file and --tls-private-key-file are provided, this flag will be ignored.
      --http2-max-streams-per-connection int
                The limit that the server gives to clients for the maximum number of streams in an HTTP/2 connection. Zero means to use golang's default.
      --permit-address-sharing
                If true, SO_REUSEADDR will be used when binding the port. This allows binding to wildcard IPs like 0.0.0.0 and specific IPs in parallel,
                and it avoids waiting for the kernel to release sockets in TIME_WAIT state. [default=false]
      --permit-port-sharing
                If true, SO_REUSEPORT will be used when binding the port, which allows more than one instance to bind on the same address and port.
                [default=false]
      --secure-port int
                The port on which to serve HTTPS with authentication and authorization. If 0, don't serve HTTPS at all. (default 10259)
      --tls-cert-file string
                File containing the default x509 Certificate for HTTPS. (CA cert, if any, concatenated after server cert). If HTTPS serving is enabled,
                and --tls-cert-file and --tls-private-key-file are not provided, a self-signed certificate and key are generated for the public address
                and saved to the directory specified by --cert-dir.
      --tls-cipher-suites strings
                Comma-separated list of cipher suites for the server. If omitted, the default Go cipher suites will be used.
                Preferred values: TLS_AES_128_GCM_SHA256, TLS_AES_256_GCM_SHA384, TLS_CHACHA20_POLY1305_SHA256, TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA,
                TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_ECDSA_WITH_AES_256_CBC_SHA, TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
                TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305, TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256, TLS_ECDHE_RSA_WITH_3DES_EDE_CBC_SHA,
                TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA,
                TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305, TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
                TLS_RSA_WITH_3DES_EDE_CBC_SHA, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_256_CBC_SHA,
                TLS_RSA_WITH_AES_256_GCM_SHA384.
                Insecure values: TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256, TLS_ECDHE_ECDSA_WITH_RC4_128_SHA, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,
                TLS_ECDHE_RSA_WITH_RC4_128_SHA, TLS_RSA_WITH_AES_128_CBC_SHA256, TLS_RSA_WITH_RC4_128_SHA.
      --tls-min-version string
                Minimum TLS version supported. Possible values: VersionTLS10, VersionTLS11, VersionTLS12, VersionTLS13
      --tls-private-key-file string
                File containing the default x509 private key matching --tls-cert-file.
      --tls-sni-cert-key namedCertKey
                A pair of x509 certificate and private key file paths, optionally suffixed with a list of domain patterns which are fully qualified domain
                names, possibly with prefixed wildcard segments. The domain patterns also allow IP addresses, but IPs should only be used if the apiserver
                has visibility to the IP address requested by a client. If no domain patterns are provided, the names of the certificate are extracted.
                Non-wildcard matches trump over wildcard matches, explicit domain patterns trump over extracted names. For multiple key/certificate pairs,
                use the --tls-sni-cert-key multiple times. Examples: "example.crt,example.key" or "foo.crt,foo.key:*.foo.com,foo.com". (default [])

Authentication flags:

      --authentication-kubeconfig string
                kubeconfig file pointing at the 'core' kubernetes server with enough rights to create tokenreviews.authentication.k8s.io. This is
                optional. If empty, all token requests are considered to be anonymous and no client CA is looked up in the cluster.
      --authentication-skip-lookup
                If false, the authentication-kubeconfig will be used to lookup missing authentication configuration from the cluster.
      --authentication-token-webhook-cache-ttl duration
                The duration to cache responses from the webhook token authenticator. (default 10s)
      --authentication-tolerate-lookup-failure
                If true, failures to look up missing authentication configuration from the cluster are not considered fatal. Note that this can result in
                authentication that treats all requests as anonymous. (default true)
      --client-ca-file string
                If set, any request presenting a client certificate signed by one of the authorities in the client-ca-file is authenticated with an
                identity corresponding to the CommonName of the client certificate.
      --requestheader-allowed-names strings
                List of client certificate common names to allow to provide usernames in headers specified by --requestheader-username-headers. If empty,
                any client certificate validated by the authorities in --requestheader-client-ca-file is allowed.
      --requestheader-client-ca-file string
                Root certificate bundle to use to verify client certificates on incoming requests before trusting usernames in headers specified by
                --requestheader-username-headers. WARNING: generally do not depend on authorization being already done for incoming requests.
      --requestheader-extra-headers-prefix strings
                List of request header prefixes to inspect. X-Remote-Extra- is suggested. (default [x-remote-extra-])
      --requestheader-group-headers strings
                List of request headers to inspect for groups. X-Remote-Group is suggested. (default [x-remote-group])
      --requestheader-username-headers strings
                List of request headers to inspect for usernames. X-Remote-User is common. (default [x-remote-user])

Authorization flags:

      --authorization-always-allow-paths strings
                A list of HTTP paths to skip during authorization, i.e. these are authorized without contacting the 'core' kubernetes server. (default
                [/healthz,/readyz,/livez])
      --authorization-kubeconfig string
                kubeconfig file pointing at the 'core' kubernetes server with enough rights to create subjectaccessreviews.authorization.k8s.io. This is
                optional. If empty, all requests not skipped by authorization are forbidden.
      --authorization-webhook-cache-authorized-ttl duration
                The duration to cache 'authorized' responses from the webhook authorizer. (default 10s)
      --authorization-webhook-cache-unauthorized-ttl duration
                The duration to cache 'unauthorized' responses from the webhook authorizer. (default 10s)

```

**Comment by johngmyers** [comment]
The question then becomes what needs to be in the SANs of these certs?

https://kubernetes.io/docs/setup/best-practices/certificates/ does not mention these server certificates.

**Comment by rifelpet** [comment]
The test in particular uses [localhost](https://github.com/kubernetes/kubernetes/pull/102050/files#diff-28a3131f96324063dd53e17270d435a3b0b3bd8f806ee0e33295929570eab209R424) and I'm guessing we can include any other local IPs matching the `--bind-address` that we end up specifying.

**Comment by olemarkus** [comment]
The self-signed certs these components provision have localhost and 127.0.0.1 in them

**Comment by olemarkus** [comment]
As mentioned on slack, the scheduler gets further than the controller-manager test.
It looks like adding this to the scheduler manifest will make the scheduler test pass:
```
    - --authentication-kubeconfig=/var/lib/kube-scheduler/kubeconfig
    - --authorization-kubeconfig=/var/lib/kube-scheduler/kubeconfig
```

The controller-manager test .. it seems like it fails to find the pod.

**Comment by johngmyers** [comment]
Since self-signed certs can't be verified, they defeat the purpose of having TLS in the first place.

Putting `localhost` and/or `127.0.0.1` in certs issued from `kubernetes-ca` would give these services the ability to impersonate kube-apiserver to other clients that run on an apiserver node.

There needs to be more thought in k/k about what these requirements are. I'm bringing the topic up in the mentioned k/k PR and in the sig-auth Slack channel.

**Comment by olemarkus** [comment]
I didn't mean to imply we should use self-signed cert. I just found that certificates is not why these tests fail.
We perhaps should split the issues in two.

The tests disable all TLS verification, and it doesn't check that `kubernetes-ca` provisined the cert, so it could be we can use a 
dedicated CA. This problem also goes for the external CCM too+, so we could end up with a larger number of certs.
